src/lib.rs

```diff
@@ -98,15 +99,125 @@ pub struct ProtectedHeader {
     pub other_fields: BTreeMap<String, String>,
 }
 
+impl ProtectedHeader {
+    /// The generation of AAD for JWE follows [A.3.5 RFC7516](https://www.rfc-editor.org/rfc/rfc7516#appendix-A.3.5)
+    pub fn generate_aad(&self) -> Vec<u8> {
+        let protected_utf8 =
+            serde_json::to_string(&self).expect("unexpected OOM when serializing ProtectedHeader");
```

Comment:
I don't think a panic in a `lib` crate is ever desired. Can we return an error here?

Reply:
Yes. I once noticed that there are no error enums in the lib, also the `serde_json::to_string` would only raise error when out of memory so panic here. Let me start with an error enum here.